include/seqan3/alphabet/composition.hpp

```diff
@@ -0,0 +1,211 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+// Author: Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+// ============================================================================
+
+#pragma once
+
+#include <cassert>
+#include <utility>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/alphabet/alphabet.hpp>
+#include <seqan3/alphabet/quality/concept.hpp>
+#include <seqan3/core/pod_tuple.hpp>
+#include <seqan3/core/detail/int_types.hpp>
+
+/*!\file alphabet/composition.hpp
+ * \ingroup alphabet
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+
+ * \brief Contains alphabet_composition.
+ */
+
+namespace seqan3
+{
+
+/*!\brief The CRTP base of alphabets that contain multiple (different) letters at one position.
+ * \ingroup alphabet
+ * \tparam first_alphabet_type Type of the first letter; must satisfy alphabet_concept.
+ * \tparam alphabet_types Types of further letters (up to 4); must satisfy alphabet_concept.
+ *
+ * This data structure is CRTP base clasee for combined alphabets, where the different
+ * alphabet letters exist independently, similar to a tuple. In fact this class
+ * provides a tuple-like interface with `std::get<0>(t)` and objects can be brace-initialized
+ * with the individual members.
+ *
+ * \attention
+ * This a "pure base class", you cannot instantiate it, you can only inherit from it.
+ * Most likely you are interested in using one of it's descendents like quality_composition.
+ *
+ * \sa quality_composition
+ * \sa mask_composition
+ */
+
+template <typename derived_type,
+          typename first_alphabet_type,
+          typename ...alphabet_types>
+      requires alphabet_concept<first_alphabet_type> && (alphabet_concept<alphabet_types> && ...)
+struct alphabet_composition :
+    public pod_tuple<first_alphabet_type, alphabet_types...>
+{
+public:
+    //!\brief The type of value_size and `alphabet_size_v<alphabet_composition<...>>`
+    using integral_type = detail::min_viable_uint_t<alphabet_size_v<first_alphabet_type> *
+                                                    (alphabet_size_v<alphabet_types> * ...)>;
+
+    //!\brief The product of the sizes of the individual alphabets.
+    static constexpr integral_type value_size{alphabet_size_v<first_alphabet_type> *
+                                              (alphabet_size_v<alphabet_types> * ...)};
+
+    //!\cond DEV
+    //!\brief the cummulative alphabet size products (first, first*second, first*second*third...) are cached
+    static constexpr std::array<integral_type, sizeof...(alphabet_types)+1> _cummulative_alph_sizes
```

Comment:
:+1: of course static private members are allowed in pod types, I had overlooked this.